department-of-veterans-affairs/caseflow
# Wmedders21/appeals 43231

Resolves [APPEALS-43231](https://jira.devops.va.gov/browse/APPEALS-43231)

# Description
Please explain the changes you made here.

- Adds new job in the `TranscriptionPackage` workflow called `Hearings::CreateBillOfMaterialsJob`
- Updates `Hearings::ZipAndUploadTranscriptionFilesJob` to avoid deleting tmp files upon completion
- Updates the factories for Hearing and LegacyHearing to have more accurate `aws_link` values for transcription files.
- Adds RSpec tests for the new job
- Adds a `json` subdirectory to `tmp/transcription_files` 

## Acceptance Criteria

## Testing Plan


1. Log in to Rails console
```
rails c -e test --sandbox
```
2. Setup data
```
hearings = (1..5).map { FactoryBot.create(:hearing, :with_transcription_files) }
legacy_hearings = (1..5).map { FactoryBot.create(:legacy_hearing, :with_transcription_files) }
def hearings_in_work_order(hearings)
  hearings.map { |hearing| { hearing_id: hearing.id, hearing_type: hearing.class.to_s } }
end
work_order = {
  work_order_name: "BVA-2030-0001",
  return_date: "xx-xx-xxxx",
  contractor_name: "Bob's contract house",
  hearings: hearings_in_work_order(hearings + legacy_hearings)
} 
```
3. Call the first two jobs in the Transcription Package workflow
```
Hearings::WorkOrderFileJob.perform_now(work_order)
Hearings::ZipAndUploadTranscriptionFilesJob.perform_now(work_order[:hearings])
```
4. Call the tested job
```
Hearings::CreateBillOfMaterialsJob.perform_now(work_order)
```
5. Expect: console output includes the following:
 `File BOM-2030-0001.json successfully uploaded to S3 location: vaec-appeals-caseflow-test/transcript_text/BOM-2030-0001.json`
6. Expect: there is now a JSON file in `<caseflow-root>/tmp/transcription_files/json named BOM-2030-0001.json`
7. Open the JSON file.
8. Audit the structure against this reference https://cyclonedx.org/docs/1.3/json/


# Frontend
## User Facing Changes

 BEFORE|AFTER
 ---|---

## Storybook Story
*For Frontend (Presentation) Components*

# Backend
## Database Changes
*Only for Schema Changes*


## Integrations: Adding endpoints for external APIs

# Best practices
## Code Documentation Updates

## Tests
### Test Coverage
Did you include any test coverage for your code? Check below:

### Code Climate
Your code does not add any new code climate offenses? If so why?

## Monitoring, Logging, Auditing, Error, and Exception Handling Checklist

### Monitoring

### Logging

### Auditing

### Error Handling

### Exception Handling
